sonar-batch/src/main/java/org/sonar/batch/cpd/AbstractCpdEngine.java

```diff
@@ -0,0 +1,109 @@
+
+import static com.google.common.collect.FluentIterable.from;
+
+public abstract class AbstractCpdEngine extends CpdEngine {
+
+  private static final Logger LOG = Loggers.get(AbstractCpdEngine.class);
+
+  static final int MAX_CLONE_GROUP_PER_FILE = 100;
+  static final int MAX_CLONE_PART_PER_GROUP = 100;
+
+  private final ReportPublisher publisher;
+  private final BatchComponentCache batchComponentCache;
+
+  public AbstractCpdEngine(ReportPublisher publisher, BatchComponentCache batchComponentCache) {
+    this.publisher = publisher;
+    this.batchComponentCache = batchComponentCache;
+  }
+
+  protected final void saveDuplications(final InputFile inputFile, List<CloneGroup> duplications) {
+    if (duplications.size() > MAX_CLONE_GROUP_PER_FILE) {
+      LOG.warn("Too many duplication groups on file " + inputFile.relativePath() + ". Keep only the first " + MAX_CLONE_GROUP_PER_FILE + " groups.");
+    }
+    final BatchComponent component = batchComponentCache.get(inputFile);
+    Iterable<org.sonar.batch.protocol.output.BatchReport.Duplication> reportDuplications = from(duplications)
+      .limit(MAX_CLONE_GROUP_PER_FILE)
+      .transform(
+        new Function<CloneGroup, BatchReport.Duplication>() {
+          private final BatchReport.Duplication.Builder dupBuilder = BatchReport.Duplication.newBuilder();
+          private final BatchReport.Duplicate.Builder blockBuilder = BatchReport.Duplicate.newBuilder();
+
+          @Override
+          public BatchReport.Duplication apply(CloneGroup input) {
+            return toReportDuplication(component, inputFile, dupBuilder, blockBuilder, input);
+          }
+
+        });
+    publisher.getWriter().writeComponentDuplications(component.batchId(), reportDuplications);
+  }
+
+  private Duplication toReportDuplication(BatchComponent component, InputFile inputFile, Duplication.Builder dupBuilder, Duplicate.Builder blockBuilder, CloneGroup input) {
+    dupBuilder.clear();
+    ClonePart originBlock = input.getOriginPart();
+    blockBuilder.clear();
+    dupBuilder.setOriginPosition(BatchReport.TextRange.newBuilder()
+      .setStartLine(originBlock.getStartLine())
+      .setEndLine(originBlock.getEndLine())
+      .build());
+    int clonePartCount = 0;
+    for (ClonePart duplicate : input.getCloneParts()) {
+      if (!duplicate.equals(originBlock)) {
+        clonePartCount++;
+        if (clonePartCount > MAX_CLONE_PART_PER_GROUP) {
+          LOG.warn("Too many duplication references on file " + inputFile.relativePath() + " for block at line " + originBlock.getStartLine() + ". Keep only the first "
+            + MAX_CLONE_PART_PER_GROUP + " references.");
+          break;
+        }
+        blockBuilder.clear();
+        String componentKey = duplicate.getResourceId();
+        if (!component.key().equals(componentKey)) {
+          BatchComponent sameProjectComponent = batchComponentCache.get(componentKey);
+          blockBuilder.setOtherFileRef(sameProjectComponent.batchId());
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'sameProjectComponent' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)